Feat/user flow: Diagrams and post on user flow in Sprout

signekb opened the pull request:
Please review the user flow post and the accompanying diagrams (which are in the post). 
These have been created based on the event storming sessions in W43 as well as talks with @K-Beicher  and @henrik-goul. 

This is meant as a first draft of concretising Sprout and what this part of Seedcase will cover, so I'm looking very much forward to your reviews :) 

Points of discussion: 

1. Schema vs table definitions
  - @K-Beicher: When we have talked about schemas, did you mean a "schema" in the sense of a "create table" script for **one** table or in the sense of a schema able to contain multiple tables, as defined below?
  - In the way we use schema is that short for *database schema*, which would be the "“blueprint” of a database which describes how the data may relate to other tables or other data models. However, the schema does not actually contain data. A sample of data from a database at a single moment in time is known as a database instance."? [source: IBM](https://www.ibm.com/topics/database-schema#:~:text=A%20database%20schema%20is%20considered,known%20as%20a%20database%20instance.)
  - L18: "The schemas in Sprout will only contain tables." Do we all agree? Or will the schemas include something else as well? Maybe sequences?

2. Headers vs. column names: What do we want to call them, @K-Beicher?

3. What the text should be on the buttons:
   - Currently, the user clicks "Save and Submit Schema" many times
   - "Upload Edited/New Version of File" - we need a better text for this, any ideas?)

4. When using file to create schema, are users able to edit headers or are they set by the data/file/csv?

Note: I have diverged from our branch strategy (GitHub Flow) by making this branch based on another feature branch, `feat/sprout-v1-diagrams`, (I made the sequence diagrams on) in an attempt to keep this PR simpler for faster reviews 🌷

signekb committed:
chore: update user flow diagram based on discussions with @K-Beicher

signekb committed:
feat: split up user flow diagram into separate diagrams for user flow post

signekb committed:
feat: add initial version of user flow post

K-Beicher commented on docs/design/images/user-flow-2-validation-schema-completeness.puml:
```diff
@@ -0,0 +1,48 @@
+:Initial schema has been created;
+  while (\nVALIDATION CHECK: Schema Completeness\nSystem checks that schema has name\nand that all headers have definitions*\n) is (Validation failed)
+    :System shows "Validation failed"
+    and s:System shows UI with schema for user to edit, 
```
s: is that a typo of wayward code?

signekb replied:
Typo. Thanks!

signekb commented:
Something we could/maybe should add to this is which user roles that can do what?

lwjohnst86 committed:
docs: :art: Use snake_case for column names, more descriptive var names, and backticks for formatting

lwjohnst86 commented on docs/design/user-flow.qmd:
```diff
@@ -0,0 +1,140 @@
+
+Together, we call these the *schema definitions*, or simply *definitions*.
+
+|schema |table | column-name | description | data-type | is-nullable | title (optional)
```
I changed this table to use more descriptive names, `snake_case` for variables, and backticks for formatting.

signekb replied:
Alrighty! So our naming convention is kebab-case for file names and snake_case for variables?

lwjohnst86 commented on docs/design/user-flow.qmd:
```diff
@@ -0,0 +1,140 @@
+
+If the user decides that the schema is not correct, they will click "Edit Schema" and they will be shown a user interface with the schema where they can edit the headers, definitions, and add new fields if needed.
+
+When the user is satisfied with their adaptions, they click "Save and Submit Schema". The schema will then go through the validation check, "Schema Completeness" again and if the validation is successful, Sprout will confirm: "New Schema added Successfully".
```
```suggestion
When the user is satisfied with their adaptions, they click "Save and Submit Schema". The schema will then go through the validation check, "Schema Completeness" again and if the validation is successful, Sprout will confirm: "New Schema Added Successfully".
```

lwjohnst86 commented on docs/design/user-flow.qmd:
```diff
@@ -0,0 +1,140 @@
+The other option is the typical [Data Contributor](users.qmd#data-contributors) route where the user will enter Sprout, click "Upload Data", and then "Use Existing Schema".
+Sprout will then show the user a list of existing schemas to choose from and the user will select a schema and select a file to upload. Then, they will click "Continue to Data Upload" (right path in the diagram above). 
+
+After these steps, Sprout will initialise the data upload. This initialisation includes a validation check, *Schema and Data Match*, as seen in the dragram below. At this step,Sprout checks whether the data and schema matches. If this validation fails, Sprout will show "Validation failed" along side error messages where the system detects that the data does not match the schema. 
```
```suggestion
After these steps, Sprout will initialise the data upload. This initialisation includes a validation check, *Schema and Data Match*, as seen in the diagram below. At this step, Sprout checks whether the data and schema match. If this validation fails, Sprout will show "Validation Failed" along with error messages where the system detects that the data does not match the schema. 
```

lwjohnst86 commented on docs/design/user-flow.qmd:
```diff
@@ -0,0 +1,140 @@
+
+After these steps, Sprout will initialise the data upload. This initialisation includes a validation check, *Schema and Data Match*, as seen in the dragram below. At this step,Sprout checks whether the data and schema matches. If this validation fails, Sprout will show "Validation failed" along side error messages where the system detects that the data does not match the schema. 
+
+The aim here is to show the users as meaningful and helpful error messages as possible. These messages should enable the user to go to their raw data and edit them to attempt to fix the errors. This step is, naturally, outside of Sprout (therefore, marked in blue in the diagram below). The user will then click "Upload Edited/New Version of File" (we need a better text for this, any ideas?) and Sp ...
```
```suggestion
The aim here is to show the users error messages that are as meaningful and helpful as possible. These messages should enable the user to go to their raw data and edit them to attempt to fix the errors. This step is, naturally, outside of Sprout (therefore, marked in blue in the diagram below). The user will then click "Upload Edited/New Version of File" and Sprout will show a user interface, where the user can add the new version of the file. When the user has selected the file and clicks "Upload File", the validation check runs again. 
```

lwjohnst86 commented on docs/design/user-flow.qmd:
```diff
@@ -0,0 +1,140 @@
+
+In the text and diagrams above, we have "chopped" the user flow of Sprout into five pieces and gone through each of them. In the diagram below, you can see the user flow in its entirety. 
+
+::: {.callout-important}
+@signekb will update entire user-flow to fit the partial user flow diagrams, when they have been reviewed/are final.
+:::
```
```suggestion
<!-- TODO: Update full user flow diagram to include all the other diagrams after review -->
```

lwjohnst86 commented on docs/design/images/user-flow-1-schema-creation.puml:
I'm not sure I understand the numbering of these files. Are they necessary/relevant? The names are all different, so the number doesn't seem necessary.

lwjohnst86 replied:
I'm also not sure I understand the colour filling choices. Why are some filled in and others aren't? If they are for things we will build sooner than later, like my previous comments, let's not go that route since the design docs are to show "what we will build and how it will look", not "what is our next plan"... since maintaining docs that are "next plan" will be ... a lot of work. Assume all design docs are (evolving) blueprints (not a roadmap).

K-Beicher replied:
The numbers denotes the flow of the files, otherwise they won't come in the correct order.  

The colour filling is a work in progress, we decided to include the V2 stuff just to make sure that we don't have people going 'but there are two things we could do here'. We could probably do a version where we just have what we are building this time, but keep these for the overall overview.

signekb replied:
@lwjohnst86 The numbering isn't strictly necessary, but it is the order they are presented in the post. It was just to make it easier for me and my tired brain to add them at the right places. I can remove the numbering. 
Yes, we implemented white backgrounds for V2 implementations to make it clear(er) (to you too) that we do not think this should be a part of the initial version of Sprout that we will aim to build during the next iteration.

signekb replied:
I'll remove the numbering of the files and the later version colouring :+1:

lwjohnst86 commented on docs/design/images/user-flow-1-schema-creation.puml:
```diff
@@ -0,0 +1,48 @@
+  :System loads file and
+  attempts to extract 
+  headers and definitions*;
+case (V2: Create Schema from\nExisting Schema)
```
```suggestion
case (Create Schema from\nExisting Schema)
```

Adding version number isn't necessary at this point, that is for us to decide during iterations. Plus, "V2" word will quickly go out of date (since we will be likely using a SemVar style of versioning (`##.##.##`).

lwjohnst86 commented on docs/design/images/user-flow-1-schema-creation.puml:
```diff
@@ -0,0 +1,48 @@
+  existing schemas;
+  #White:User inspects schemas
+  and selects base schema;
+case (V3+: Create Schema Manually)
```
```suggestion
case (Create Schema Manually)
```

lwjohnst86 commented on docs/design/images/user-flow-1-schema-creation.puml:
```diff
@@ -0,0 +1,48 @@
+title User Flow: Schema Creation
+
+:User selects option to "Upload Data";
+'Data vs file: File indicates only a single file (e.g., csv) can be uploaded at a time. Is that what we want?
```
```suggestion
```

Moving this comment out to here. Good question... I don't see why we couldn't upload multiple files that are each scanned and create multiple schemas from them. Maybe for now we assume multiple (without explicitly saying it here?).

K-Beicher replied:
I don't think I've ever seen a system which allowed for multiple files to be uploaded at the same time if they require the level of validation that we suggest will take place.  Also, can you imagine the hassle of attaching several files just to have to do it again because one of them failed.  We may want to add this as a discussion item to one of our meetings.

signekb replied:
I'll create an issue on it 👍

lwjohnst86 commented on docs/design/images/user-flow-1-schema-creation.puml:
```diff
@@ -0,0 +1,48 @@
+
+:User selects option to "Upload Data";
+'Data vs file: File indicates only a single file (e.g., csv) can be uploaded at a time. Is that what we want?
+'That could be very annoying for researchers with one file per participant (per visit)
```
```suggestion
```

lwjohnst86 commented on docs/design/images/user-flow-1-schema-creation.puml:
```diff
@@ -0,0 +1,48 @@
+  clicks "Create Schema";
+  :System loads file and
+  attempts to extract 
+  headers and definitions*;
```
What's the `*` for?

K-Beicher replied:
There is a footnote in the full overview which explains what the 'definitions' cover.  The footnote should be brought over, we are aware of this.

signekb replied:
I have moved the footnote so it's in the diagram as well.

lwjohnst86 commented on docs/design/images/user-flow-2-validation-schema-completeness.puml:
```diff
@@ -0,0 +1,48 @@
+    - option to "Add field"
+    Missing fields are highlighted;
+    :User completes/edits schema;
+    if (V2: System asks user if they want to save only or submit) then (V2: Save only)
```
```suggestion
    if (System asks user if they want to save only or submit) then (Save only)
```

lwjohnst86 commented on docs/design/images/user-flow-2-validation-schema-completeness.puml:
```diff
@@ -0,0 +1,48 @@
+    including: 
+    - headers (textbox)
+    - textbox for description (textbox)
+    - data type (dropdown?)
```
```suggestion
    - data type (dropdown)
```

lwjohnst86 commented on docs/design/images/user-flow-2-validation-schema-completeness.puml:
```diff
@@ -0,0 +1,48 @@
+    including: 
+    - headers (textbox)
+    - textbox for description (textbox)
+    - data type (dropdown?)
```
```suggestion
    - data type (dropdown)
```

lwjohnst86 commented on docs/design/images/user-flow-2-validation-schema-completeness.puml:
```diff
@@ -0,0 +1,48 @@
+:System shows UI with schema for user confirmation;
+  while (User decides whether schema is correct or needs editing) is (Incorrect schema)
+    :User clicks "Edit Schema";
+    :System shows UI with schema for user to edit, 
+    including: 
```
```suggestion
    :System shows UI with schema for 
    user to edit, including: 
```

lwjohnst86 commented on docs/design/images/user-flow-2-validation-schema-completeness.puml:
```diff
@@ -0,0 +1,48 @@
+:Initial schema has been created;
+  while (\nVALIDATION CHECK: Schema Completeness\nSystem checks that schema has name\nand that all headers have definitions*\n) is (Validation failed)
+    :System shows "Validation failed"
+    and s:System shows UI with schema for user to edit, 
+    including: 
```
```suggestion
    and shows UI with schema for 
    user to edit, including: 
```

lwjohnst86 commented on docs/design/images/user-flow-2-validation-schema-completeness.puml:
```diff
@@ -0,0 +1,48 @@
+    - option to "Add field";
+    :User edits schema;
+    :User clicks "Continue";
+'Arrow from "User edits definitions" should go back up to VALIDATION CHECK      
+    :Arrow that goes to "User decides whether schema is correct [...]"
+    should go back to Validation check
+    System goes back to VALIDATION CHECK: Schema Completeness;
```
Ah, that's what you were frustrated with... Yea, that's weird... there's got to be a way to fix that. Could it end at "Restart with uploading validation check?"

K-Beicher replied:
We've posted it to the Forums on plantUML, we'll see if anyone gets back to us, otherwise that is an option.

signekb replied:
We've got a response, with a not so pretty but better, workaround. I have implemented that now 🎉

lwjohnst86 commented on docs/design/images/user-flow-3-schema-created-continue-to-upload.puml:
```diff
@@ -0,0 +1,16 @@
+
+:...;
+:User clicks "Save and Submit Schema";
+#palegreen:Systems confirms: "New Schema added Successfully";
```
```suggestion
#palegreen:System confirms: "New Schema added Successfully";
```

lwjohnst86 commented on docs/design/images/user-flow-4-prepare-to-upload-using-existing-schema.puml:
```diff
@@ -0,0 +1,22 @@
+title User Flow: Prepare to Upload using Existing Schema
+
+:User selects option to "Upload Data";
+'Data vs file: File indicates only a single file (e.g., csv) can be uploaded at a time. Is that what we want?
+'That could be very annoying for researchers with one file per participant (per visit)
```
```suggestion
```
See comment above about this.

lwjohnst86 commented on docs/design/images/user-flow-5-data-upload.puml:
```diff
@@ -0,0 +1,21 @@
+title Data Upload
+
+:...;
+:System initialises data upload;
+while (\nVALIDATION CHECK: Schema and data match\nSystem checks if data and schema matches\n) is (Validation failed)
```
```suggestion
:System initialises data upload;
while (\nVALIDATION CHECK: Check if Schema and data match\n) is (Validation failed)
```

lwjohnst86 commented on docs/design/images/user-flow-5-data-upload.puml:
```diff
@@ -0,0 +1,21 @@
+:...;
+:System initialises data upload;
+while (\nVALIDATION CHECK: Schema and data match\nSystem checks if data and schema matches\n) is (Validation failed)
+:Systems shows "Upload failed" along\nwith description on where data and\nschema do not match;
```
```suggestion
:Systems shows "Upload Failed" along\nwith description on where data and\nschema do not match;
```

lwjohnst86 reviewed (changes requested):
Really great job you two on mapping this out! I've left several comments and suggested changes. I think it describes the flow very well!!! :clinking_glasses: :raised_hands: :tada:

signekb commented on docs/design/user-flow.qmd:
```diff
@@ -0,0 +1,140 @@
+
+NOTE: ADD USER ROLES MORE SPECIFICALLY THROUGHOUT THIS.
+
+::: {.callout-important}
+Things we need to define/get clearer on: 
+
+- Schema vs table definitions
+  - @K-Beicher: When we have talked about schemas, did you mean a "schema" in the sense of a "create table" script for **one** table or in the sense of a schema able to contain multiple tables, as defined below?
+  - In the way we use schema is that short for *database schema*, which would be the "“blueprint” of a database which describes how the data may relate to other tables or other data models. However, the schema does not actually contain data. A sample of data from a database at a single moment in time is known as a database instance."? [source: IBM](https://www.ibm.com/topics/database-schema#:~:te ...
+- Data vs file upload? Is it only a file or also multiple files at a time? 
+- What the text should be on the buttons (currently, the user clicks "Save and Submit Schema" many times)
+- ... other things?
+
+:::
```
Duly noted. I'll remove it here and add it to the PR

signekb committed:
chore: renmae user flow files and minor edits based on review from @lwjohnst86

signekb committed:
fix: add line from "continue" back validation check

Hotfix. The line is red and not very neat, but still an improvement from not having it

signekb committed:
chore: Redo title to avoid /

K-Beicher commented on docs/design/images/user-flow-5-data-upload.puml:
```diff
@@ -0,0 +1,21 @@
+    :User selects file and clicks "Upload File";
+
+end while (Validation successful);
+#palegreen:System displays confirmation: "Upload Successful";
```
We should probably have a couple of boxes more here.  One to say that the system displays back to the user the number of rows it intends to write to the database, with an option to abort if the number isn't what they expected and another with upload successful.

So something like this between the last two lines of code:
while (System displays the number of \nsuccessfully validated rows \nto the user for agreement) is (No)
:User clicks "Cancel Upload";
end
end while (Yes);

signekb committed:
docs: text mods based on review by @lwjohnst86

signekb committed:
docs: Minor text mods from readthrough

signekb committed:
docs: Add Data Access Administrator throughout text instead of user

+ remove blank lines in the bottom of the post

signekb committed:
fix: Minor fixes to user flow diagrams based on review from @lwjohnst86

signekb committed:
feat: add "system displays the number of validated rows to the user for agreement"

From review by @K-Beicher

signekb committed:
fix: remove numbering from output file name

signekb committed:
chore: :sparkles: add pngs from updated pumls

signekb committed:
fix: :fire: remove "tricky part" png

this was a draft of an attempt of creating the arrow back to validation

signekb committed:
fix: :fire: delete outdated schema creation flow diagram

signekb committed:
fix: remove mention of "V2"

signekb committed:
feat: add user confirmation of validated row to add in upload

signekb committed:
style: change variable names in footer to lower case

signekb committed:
refactor: :sparkles: update to match partial user flow diagrams

signekb commented:
@lwjohnst86 I believe that I have implemented the changes you suggested, but because I did it through VS Code, it seems like GitHub hasn't "recorded" it?

lwjohnst86 commented on docs/design/user-flow.qmd:
```diff
@@ -0,0 +1,118 @@
+
+{{< include ../_wip.qmd >}}
+
+This document describes the user flow of Sprout. The user flow is split into five diagrams that go through the flow in more detail. At the end of the post, you can see the user flow of Sprout in its entirety.
```
```suggestion
This document describes the user flow of Sprout. The user flow is split into five diagrams that go through the flow in more detail. At the end of the document, you can see the user flow of Sprout in its entirety.
```

lwjohnst86 commented on docs/design/user-flow.qmd:
```diff
@@ -0,0 +1,118 @@
+| `data_resource` | `table_1` | `email` | Unique ID identifying the user | `int` | 0 | NULL |
+| `data_resource` | `table_1` | `address` | first address line of user's address | `varchar` | 1 | address line 1 |
+
+: Example of schema named `data resource` with one table `table_1` with user information
```
```suggestion
: Example of schema named `data resource` with one table `table_1` with user information.
```

lwjohnst86 commented on docs/design/user-flow.qmd:
```diff
@@ -0,0 +1,118 @@
+- Create Schema from Existing Schema
+- Create Schema Manually
+
+![](images/user-flow-schema-creation.png)
```
```suggestion
![Diagram of the flow that occurs when a new schema is created.](images/user-flow-schema-creation.png)
```

lwjohnst86 commented on docs/design/user-flow.qmd:
```diff
@@ -0,0 +1,118 @@
+
+Now, that the schema has been created successfully, the [Data Management Administrator](users.qmd#data-management-administrators) has two options: Either they can "Continue to Data Upload", if they want to add data to the newly created schema right after the schema has been submitted, or they can choose to quit Sprout at this step, if they only wanted to create the schema as preparation for futur ...
+
+![](images/user-flow-schema-created-continue-to-upload.png)
```
```suggestion
![Diagram describing the next steps given after creating the schema and passing the validation steps.](images/user-flow-schema-created-continue-to-upload.png)
```

lwjohnst86 commented on docs/design/user-flow.qmd:
```diff
@@ -0,0 +1,118 @@
+
+There are two entry points for a user to upload data using an existing schema. The first option is to continue from the diagram above, where the [Data Management Administrator](users.qmd#data-management-administrators) will upload data right after they have created a new schema. In this option, they click "Continue to Data Upload" (left path of diagram below).
+
+![](images/user-flow-prepare-to-upload-using-existing-schema.png)
```
```suggestion
![Diagram showing the steps taken for uploading data when using an existing schema.](images/user-flow-prepare-to-upload-using-existing-schema.png)
```

lwjohnst86 commented on docs/design/user-flow.qmd:
```diff
@@ -0,0 +1,118 @@
+
+If validation succeeds, the data will be uploaded and Sprout will confirm: "Upload Successful". 
+
+![](images/user-flow-data-upload.png)
```
```suggestion
![Diagram showing the steps that occur after validation and final data upload from a Data Contributor.](images/user-flow-data-upload.png)
```

lwjohnst86 commented on docs/design/user-flow.qmd:
```diff
@@ -0,0 +1,118 @@
+
+<!-- TODO: Update full user flow diagram to include all the other diagrams after review -->  
+
+![](images/user-flow.png)
```
```suggestion
![Diagram showing the full flow from user inputting data into it finally being stored in the database.](images/user-flow.png)
```

lwjohnst86 committed:
fix: apply suggestions from code review

lwjohnst86 commented:
Looks fabulous!
